### ISSUE TITLE
HMR for Angular RC5 and beyond

### ISSUE THREAD

**Issue by MikeRyan52** [CodeSequence/ngrx-store-hmr#2]
See here for initial work towards this goal: https://github.com/teropa/ngrx-hotload-rc5

I'm wondering if we could wrap `platformBrowserDynamic` somehow to provide a good development-only API for Store and HMR.

**Comment by teropa** [comment]
@MikeRyan52 I think that would be ideal, but there's a couple of issues I ran into when I tried to do this by bypassing `NgModule`. My attempt is in this branch: https://github.com/teropa/ngrx-hotload-rc5/tree/provider-bootstrap

I tried to pass a Store provider as part of `bootstrapModule` [here](https://github.com/teropa/ngrx-hotload-rc5/blob/provider-bootstrap/src/main.ts#L8-L11) instead of using an `NgModule`. This would be very close to the pre-rc.5 solution. But I'm getting "No provider for Store!" exceptions from this, so it doesn't seem like the provider is there at runtime. Not sure what's going on there.

The other issues is that the `provideStore` function is now [deprecated](https://github.com/ngrx/store/blob/master/src/ng2.ts#L40-L54) which leads me to think we shouldn't be using it. I guess if the first issue could be resolved maybe this could be undeprecated though.. :)

**Comment by teropa** [comment]
FYI: Ran into an issue with passing the store provider into `platformBrowserDynamic`: The store then goes into the platform injector which is only ever created once. So with this approach *you can't hot reload reducer code changes*.

That's a no-go for me, so in my app I reverted back to my first approach (the one described in https://github.com/teropa/ngrx-hotload-rc5)

**Comment by DominikDitoIvosevic** [comment]
For someone not intimately familiar with the current development of this, is there a plug and play version for post rc5 or maybe a different lib for this?

**Comment by teropa** [comment]
@DominikDitoIvosevic No plug and play version at the moment, I'm afraid. The closest thing I'm aware of is the setup in https://github.com/teropa/ngrx-hotload-rc5.

There's also https://github.com/AngularClass/angular2-hmr but that's also more like infrastructure than a plug and play thing you can just add to your app.

**Comment by MikeRyan52** [comment]
@teropa I've been playing around with the example from your harmonics-explorer app and think I have a pretty decent solution put together.

Code:
```ts
import { Provider, PlatformRef, Type, NgModuleRef, ApplicationRef } from '@angular/core';
import { _INITIAL_STATE, Store } from '@ngrx/store';
import { take } from 'rxjs/operator/take';
import { publishReplay } from 'rxjs/operator/publishReplay';
import { mergeAll } from 'rxjs/operator/mergeAll';
import { Observable } from 'rxjs/Observable';
import { Observer } from 'rxjs/Observer';

export type PlatformFactoryFn = (extraProviders?: Provider[]) => PlatformRef;


export function useHMR<T>(module: any, platform: PlatformFactoryFn, ngModule: Type<T>): Observable<NgModuleRef<T>> {
  const bootstrap$ = new Observable((observer: Observer<Promise<NgModuleRef<T>>>) => {
    let state: any;
    let moduleRef: NgModuleRef<T>;

    const bootstrap = () =>  {
      const stateProvider: Provider = { provide: _INITIAL_STATE, useValue: state };

      observer.next(
        platform([ stateProvider ]).bootstrapModule(ngModule)
          .then(ref => moduleRef = ref)
      );
    };

    if (document.readyState === 'complete') {
      bootstrap();
    }
    else {
      document.addEventListener('DOMContentLoaded', bootstrap);
    }

    module.hot.accept();

    module.hot.dispose(() => {
      const appRef: ApplicationRef = moduleRef.injector.get(ApplicationRef);
      const store: Store<any> = moduleRef.injector.get(Store);

      take.call(store, 1).subscribe((s: any) => state = s);

      appRef.components.forEach(cmp => {
        const node = cmp.location.nativeElement;

        const newNode = document.createElement(node.tagName);
        node.parentNode.insertBefore(newNode, node);
      });

      moduleRef.destroy();
    });
  });

  const merged$ = mergeAll.call(bootstrap$);
  const published$ = publishReplay.call(merged$, 1).refCount();

  published$.subscribe();

  return published$;
}
```

Usage:
```ts
import './polyfills';
import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
import { AppDevModule } from './module.dev';
import { useHMR } from './hmr';

declare var module: any;

useHMR(module, platformBrowserDynamic, AppDevModule);
```

Very minimal setup for the end user. Thoughts?

**Comment by teropa** [comment]
@MikeRyan52 I like the end user ergonomics!

I tried this with a project of mine but I'm not getting my `_INITIAL_STATE` factory invoked at all. It seems to me that `StoreModule` is using its own provider for that token instead. Is there a change needed for the way Store setup is done?

**Comment by firecube** [comment]
@MikeRyan52, @teropa  agreed, I like the look of this minimal setup as I have struggled getting my nested state restored.  Unfortunately this doesn't work for me either.

**Comment by qdouble** [comment]
@MikeRyan52 any progress with this? I'm looking to add HMR but not sure I should wait on this or use AngularClass.

**Comment by Mark-McCracken** [comment]
Just wondering if anyone still has any plans to work on this? I know it was a while ago, but I've been trying unsuccessfully to get hmr to preserve state using a cli generated project.
Was wondering if anyone in here had managed to do just that?

**Comment by kirillgroshkov** [comment]
+1
